### ISSUE TITLE
R2RMLTC0002f

### ISSUE THREAD

**Issue by chrdebru** [kg-construct/r2rml-test-cases-support#7]
Is it possible to clarify this test? The purpose is: "Tests the presence of delimited identifiers referenced as regular identifiers. Within rr:template ID is ok, but Name is not". Using delimited variables in rr:column is fine [1].

I see: rr:column "\\"ID\\"" and rr:column "\\"Name\\"" 

Delimited identifiers are case sensitive [1] and Name does correspond with 

CREATE TABLE "Student" ("ID" INTEGER, "Name" VARCHAR(15)); (d0002.sql)

[1] https://www.w3.org/TR/r2rml/#logical-tables

Either there is a mistake in the test, or the message and purpose of the test are not clear.

EDIT: Had to update the issue as backslashes were unescaped.

**Comment by dachafra** [comment]
I think the test is for  `rr:template "http://example.com/{ID}/{Name}"` in the SubjectMap as ID and "ID" are the same but Name should be \"Name\"

**Comment by chrdebru** [comment]
Name in the template is a regular identifier. Regular identifiers are case-insensitive and thus corresponds with Name, name, NAME, etc. It happens to be equal to the column name Name. The delimited identifier in the column name is case-sensitive and corresponds with the name of the column. Both column identifiers in the template are valid identifiers.

**Comment by ArenasGuerreroJulian** [comment]
R2RML spec: "delimited identifiers that are not all-upper-case are not equivalent to any undelimited identifiers". For me this is a contradiction in the spec and may create confusion in this test case.

The reference `Name` that is in the subject map is not delimited and is then treated as `NAME`, which is not correct since the in the DB table it is `Name`.

**A potentially pollemic note**: I do not see the need for regular and delimited identifiers, for me everything should be treated as a deliited identifier.

**Comment by chrdebru** [comment]
My understanding was that regular identifiers are case insensitive. So they should not be treated as `NAME`. In the test case, the logical table uses an `rr:tableName`, which yields `SELECT * FROM \\"Students\\".` The columns of the executed query are not (necessarily) delimited.

**Comment by jatoledo** [comment]
Hi @chrdebru, Thanks for this good question.


Everything depends on what your schema represents in the database. In our case ([D002-1table2columns1row](https://www.w3.org/2001/sw/rdb2rdf/test-cases/#D002-1table2columns1row)) we want to represent the column "Name" (not a regular identifier).
Note: Read comments in the code.

[D002-1table2columns1row](https://www.w3.org/2001/sw/rdb2rdf/test-cases/#D002-1table2columns1row)
```sql
CREATE TABLE "Student" (
"ID" INTEGER,
"Name" VARCHAR(15)    ---> Here we represent "Name", not a regular identifier.
                      ---> You can also try to experiment with: Name VARCHAR(15) 
);
INSERT INTO "Student" ("ID", "Name") VALUES(10,'Venus');
```
So, in the mapping, you write the correct identifier:

```yml
<TriplesMap1>
    a rr:TriplesMap;
    
    rr:logicalTable [ rr:tableName "\"Student\"" ];

    rr:subjectMap [ rr:template "http://example.com/{ID}/{Name}";    #----->Here: Name represent a regular identifier
                                         # The correct is : \"Name\", because is this that represent in the database
                    rr:class foaf:Person ]; 

    rr:predicateObjectMap
    [ 
      rr:predicate		ex:id ; 
      rr:objectMap		[ rr:column "\"ID\"";  ]
    ];

    rr:predicateObjectMap
    [ 
      rr:predicate		foaf:name ; 
      rr:objectMap		[ rr:column "\"Name\"" ]    #------------>Here: "Name" is not a regular identifier
].
```

Then, what happens with the databases.. Here a short example:
```sql

/********************************* Good for both *****************************/
/*******************\
MySQL
\*******************/
mysql> select * from "Student";
+------+-------+
| ID   | Name  |
+------+-------+
|   10 | Venus |
+------+-------+
/*******************\
PostgreSQL
\*******************/
r2rml=# select * from "Student";
 ID | Name
----+-------
 10 | Venus

/*******************\
MySQL
\*******************/
mysql> select "Name" from "Student";
+-------+
| Name  |
+-------+
| Venus |
+-------+

/*******************\
PostgreSQL
\*******************/
r2rml=# select "Name" from "Student";
 Name
-------
 Venus


/********************************* Oh oh, we have a problem ¯\(°_o)/¯ *****************************/
/*******************\
MySQL
\*******************/
mysql> select Name from "Student";
+-------+
| Name  |
+-------+
| Venus |
+-------+
/*******************\
PostgreSQL
\*******************/
r2rml=# select Name from "Student";
ERROR:  column "name" does not exist
LINE 1: select Name from "Student";


```
Finally, my question is: If I put only Name in the mapping, who solves the problem? , the engine, the databases??. IMO, writing the identifier correctly in the mapping solves this problem.

Thank you again for you question.

**Comment by chrdebru** [comment]
The problem is thus that MySQL does not comply with ANSI SQL w.r.t. to this particular issue. As you cannot derive from the mapping that the column `"Name"` was created for the table. I have no access to a PostgreSQL instance on this machine, but one could create a table with `Name `and `"Name"` as two columns, no? To add to your questions: would it be up to the engine to analyze the logical tables prior to executing the mapping? E.g., by analyzing the information schema. Is that part of the mapping engine's concern? I agree that you can solve this by placing the correct identifier in mapping. The test executed for MySQL could come with a comment; the test still fails, but could have passed if an implementation does the necessary checks.

**Comment by jatoledo** [comment]
In these test-cases you can see/know the schema SQL(are public). About your question, it could create a table with Name and "Name" as two columns in PostgreSQL, the answer is yes, but I would like to clarify that again it depends on the database. 
Writing in PostgreSQL
 ```bash
Name --> name
"Name" --> Name
*** Are different
```
Writing in Mysql
```bash
 Name --> Name
"Name" --> Name
** Errror: Duplicate column
```

But, this is not the problem(it's another topic).  From my point of view, the mapping should be more uncoupled as possible from the databases, the real problem is when users/developers work with only one database. For example Mysql/MariaDB, in this case 
```  rr:template "http://example.com/{ID}/{Name}" ``` and ```  rr:template "http://example.com/{ID}/{\"Name\"}"-->(R2RML TEST CASES) ``` works. So, why do we force you to put double quotation marks if the first example works? If the first one works for them, I doubt they will implement the second case. Maybe it would be a solution to modify the mapping for Mysql (transparent to the user) in order to do more usable. With PostgreSQL, we don't have any problem(I don't remember very well but if I'm not wrong Oracle works the same). What do you think?

**Comment by dachafra** [comment]
Closing issue, it will be solved in #14
